### PR TITLE
introduce tekton controller

### PR DESCRIPTION
**Pull request by dipak-pawar**
*Changes Propose in this PR*
- Introduce Tekton Controller with TektonInstallation Custom Resource.
- Moved all logic belongs to Tekton from Che controller to Tekton

Note: I'll be refactoring che controller and it's tests in next PR, so you can made it less focused in your review. I just want to fix all failing tests for che due to this refactoring. You should expect PR once this gets merged

Fix: https://jira.coreos.com/browse/CRT-325

### CODE REVIEW

**Comment by MatousJobanek** [comment]
@dipak-pawar Could you please split the PR:
* one for the changes in api / CRD
* one for the controller logic

thanks

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_tektoninstallation")
+
+// Add creates a new TektonInstallation Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	cr := toolchain.NewTektonInstallation()
+	if err := mgr.GetClient().Create(context.TODO(), cr); err != nil && !errors.IsAlreadyExists(err) {
```
Comment (MatousJobanek) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: I thought that this should be in a separated PR - please keep the PRs as small as possible

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 2]
```diff
@@ -85,23 +83,12 @@ func (r *ReconcileCheInstallation) Reconcile(request reconcile.Request) (reconci
...
-		errors = append(errors, err)
-	} else if isResourceCreated {
-		return reconcile.Result{Requeue: true}, nil
-	}
-
-	if isResourceCreated, err := r.EnsureTektonSubscription(reqLogger, cheInstallation); err != nil {
-		errors = append(errors, err)
+		return reconcile.Result{}, err
 	} else if isResourceCreated {
 		return reconcile.Result{Requeue: true}, nil
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: The main reason for doing it differently here was the fact that we need to create the namespace only once when we installing the operator. Having a permanent watcher may be an overkill in this case. In member operator it's different. It's not one time action there.
On the other hand if the namespace get deleted then we won't restore it automatically without a watcher.
Anyway, I think it's OK if we keep it as is for now in this PR because it's about moving existing logic to a different controller. And we can modify the logic itself in a separate PR.

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 1 above)
Comment (alexeykazakov) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: I agree. Let's do the TektonInstallation reconcile logic from CheInstallation controller to TektonInstallation controller in one PR (this one) and creating Installations automaticaly during operator startup in a separate PR.

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: Yes, This is what my plan is to have PR small

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+	if err := r.EnsureTektonSubscription(reqLogger, tektonInstallation); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileTektonInstallation) EnsureTektonSubscription(logger logr.Logger, tektonInstallation *v1alpha1.TektonInstallation) error {
+	tektonSubNamespace := tekton.SubscriptionNamespace
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Yes, we are 100% sure that, it already exists

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	tektonInstallation := &toolchainv1alpha1.TektonInstallation{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: toolchain.TektonInstallation}, tektonInstallation); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, maybe someone deleted it, but we need this to be exist, so creating it.
+			cr := toolchain.NewTektonInstallation()
+			if err := r.client.Create(context.TODO(), cr); err != nil {
+				reqLogger.Error(err, "failed to create tekton installation custom resource", "Namespace", request.Namespace, "Name", cr.Name)
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{Requeue: true}, nil
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Good point, we can get rid out of `reconcile.Result{Requeue: true}`.

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 1 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: sure, let me make it as a part of separate PR

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go [hunk 5]
```diff
@@ -158,24 +120,18 @@ func TestCheInstallationController(t *testing.T) {
...
 		// given
 		cheOperatorNs, cheOg, cheSub := newCheResources()
-		tektonSub := tekton.NewSubscription(tekton.SubscriptionNamespace)
-
-		cl, r := configureClient(t, cheOperatorNs)
-
 		cheInstallation := NewCheInstallation(cheOperatorNs)
+		cl, r := configureClient(t)
+
 		request := newReconcileRequest(cheInstallation)
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: I believe, we are just testing that we have handled a case `reconcile is not triggered with resource is not created/present`

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go [hunk 6]
```diff
@@ -105,36 +97,6 @@ func TestCheInstallationController(t *testing.T) {
...
-			AssertThatSubscription(t, cheSub.Namespace, cheSub.Name, cl).
-				Exists().
-				HasSpec(cheSub.Spec)
-
-			AssertThatSubscription(t, tektonSub.Namespace, tektonSub.Name, cl).
-				Exists().
-				HasSpec(tektonSub.Spec)
 		})
 
 		t.Run("should not requeue", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: sure, that's true, But I believe, it's good to have this kind of tests to make sure that everything works correctly, instead of depending on e2e tests to debug/fix any issue

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go [hunk 7]
```diff
@@ -261,11 +203,6 @@ func TestCheInstallationController(t *testing.T) {
 			require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: I don't understand what do you want to suggest here. Could you please elaborate more?

File: pkg/controller/tektoninstallation/tektoninstallation_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"testing"
+)
+
+func TestTektonInstallationController(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller_test.go]: I wants to keep it under separated group using subtests

File: test/e2e/toolchain_test.go [hunk 9]
```diff
@@ -45,10 +47,7 @@ func TestToolchain(t *testing.T) {
 		// then
```
Comment (nurali-techie) [inline test/e2e/toolchain_test.go]: Minor: `then` is mis-placed

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling TektonInstallation")
+
+	// Fetch the TektonInstallation instance
+	tektonInstallation := &toolchainv1alpha1.TektonInstallation{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: toolchain.TektonInstallation}, tektonInstallation); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, maybe someone deleted it, but we need this to be exist, so creating it.
+			cr := toolchain.NewTektonInstallation()
+			if err := r.client.Create(context.TODO(), cr); err != nil {
```
Comment (nurali-techie) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Generally we avoid creating primary_object in controller which managing the object.  Here we are creating `TektonInstallation` in it's own controller.  Hope this point is already considered.  Can you plz clarify reasoning behind it.

File: test/e2e/toolchain_test.go (hunk 9 above)
Comment (dipak-pawar) [reply test/e2e/toolchain_test.go]: not sure how is it misplaced?

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 10 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: That's correct @nurali-techie, but our requirement here is to create required customresources like   `CheInstallation/TektonInstallation` automatically by our operator, we don't want user to create this custom resources, as this is simplified opinionated flow for user. This is why we are created it upfront before starting controller and also in reconcile to handle use case if user accidentally deletes `TektonInstallation`

File: test/e2e/toolchain_test.go (hunk 9 above)
Comment (nurali-techie) [reply test/e2e/toolchain_test.go]: sorry .. I got it wrong .. it's fine :+1:

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 10 above)
Comment (nurali-techie) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: @dipak-pawar thanks for clarification.

File: .codecov.yaml [hunk 11]
```diff
@@ -28,7 +28,7 @@ coverage:
 # See http://docs.codecov.io/docs/pull-request-comments-1
 comment:
   layout: "header, diff, tree"
-  behavior: # default = posts once then update, posts new if delete
+  #behavior: # default = posts once then update, posts new if delete
```
Comment (nurali-techie) [inline .codecov.yaml]: any reason behind disabling ??

File: pkg/apis/toolchain/v1alpha1/tektoninstallation_types.go [hunk 12]
```diff
@@ -21,6 +22,14 @@ type TektonInstallationStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
 	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// Conditions is an array of current TektonInstallation conditions
+	// Supported condition types:
+	// TektonInstalled, FailedToInstallTekton
```
Comment (nurali-techie) [inline pkg/apis/toolchain/v1alpha1/tektoninstallation_types.go]: Need to correct type name -> Installed, FailedToInstall

File: .codecov.yaml (hunk 11 above)
Comment (dipak-pawar) [reply .codecov.yaml]: lint check was failing.

File: pkg/apis/toolchain/v1alpha1/tektoninstallation_types.go (hunk 12 above)
Comment (nurali-techie) [reply pkg/apis/toolchain/v1alpha1/tektoninstallation_types.go]: Even type is one `TektonReady` and Installed, FailedToInstall are reason

File: pkg/controller/add_tektoninstallation.go [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package controller
+
+import (
+	"github.com/codeready-toolchain/toolchain-operator/pkg/controller/tektoninstallation"
+)
+
+func init() {
+	// AddToManagerFuncs is a list of functions to create controllers and add them to a manager.
+	AddToManagerFuncs = append(AddToManagerFuncs, tektoninstallation.Add)
+}
```
Comment (nurali-techie) [inline pkg/controller/add_tektoninstallation.go]: Suggestion: We can add directly into `controller.go` rather adding new file for each controller with one_line.  Example: https://github.com/nurali-techie/member-operator/blob/master/pkg/controller/controller.go

Comment (dipak-pawar) [reply pkg/controller/add_tektoninstallation.go]: Addressed in  64ab1f5

File: pkg/apis/toolchain/v1alpha1/tektoninstallation_types.go (hunk 12 above)
Comment (dipak-pawar) [reply pkg/apis/toolchain/v1alpha1/tektoninstallation_types.go]: addressed in c3c3805

File: deploy/crds/toolchain.openshift.dev_tektoninstallations_crd.yaml [hunk 14]
```diff
@@ -33,6 +33,35 @@ spec:
           type: object
         status:
           description: TektonInstallationStatus defines the observed state of TektonInstallation
+          properties:
+            conditions:
+              description: 'Conditions is an array of current TektonInstallation conditions
+                Supported condition types: TektonInstalled, FailedToInstallTekton'
```
Comment (nurali-techie) [inline deploy/crds/toolchain.openshift.dev_tektoninstallations_crd.yaml]: I guess, after change of condition type, we need to run generate to correct this too.

Comment (dipak-pawar) [reply deploy/crds/toolchain.openshift.dev_tektoninstallations_crd.yaml]: addressed in ba76ab5

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+// and what is in the TektonInstallation.Spec
+func (r *ReconcileTektonInstallation) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling TektonInstallation")
+
+	// Fetch the TektonInstallation instance
+	tektonInstallation := &toolchainv1alpha1.TektonInstallation{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: toolchain.TektonInstallation}, tektonInstallation); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
```
Comment (nurali-techie) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Now I see you removed creation here, what's the reason ??

Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: we'll address it in another PR, create issue for it https://jira.coreos.com/browse/CRT-344

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+func (r *ReconcileTektonInstallation) EnsureTektonSubscription(logger logr.Logger, tektonInstallation *v1alpha1.TektonInstallation) error {
+	tektonSubNamespace := tekton.SubscriptionNamespace
+	if err := r.ensureTektonSubscription(logger, tektonSubNamespace, tektonInstallation); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, tektonInstallation, r.setStatusTektonSubscriptionFailed, err, "failed to create tekton subscription in namespace %s", tektonSubNamespace)
+	}
+	return r.StatusUpdate(logger, tektonInstallation, r.setStatusTektonSubscriptionReady, tekton.SubscriptionSuccess)
+}
+
+func (r *ReconcileTektonInstallation) ensureTektonSubscription(logger logr.Logger, ns string, tektonInstallation *v1alpha1.TektonInstallation) error {
```
Comment (nurali-techie) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Minor: for consistency, we can move `ns` from place 2 to 3 in input args.

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+	tektonSub := tekton.NewSubscription(ns)
```
Comment (nurali-techie) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: It would be good if we call `tekton.NewSubscription(ns)` when Get() failed with NotFound and we are actually going to create object.

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: I didn't want to say that we should depend only on e2e tests, the unit tests are needed for sure, but this is not really a unit test. Instead of depending on the previous subtests it should contain a proper test setup so the test is isolated and can be executed on its own. 
In case you don't want to do the proper setup for every subtest separately, then another option would be embedding the subtests into the one it depends on - in other words, the `"should not requeue"` would be a subtest of `"should create che subscription and requeue"` and that would be a subtest of `"should create operator group and requeue"` and so on - to make it clear that they depend on the previous ones

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 17 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Addressed in 31d139b

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 16 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Addressed in 31d139b

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: but how the reconcile could be triggered for a resource that is not created/present? I'm probably missing something

[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: If resource is not exist/present, it won't trigger reconcile, but we need to cover this to improve code-coverage and handle following scenario
```
if err := r.client.Get(context.TODO(), types.NamespacedName{Name: toolchain.TektonInstallation}, tektonInstallation); err != nil {
		if errors.IsNotFound(err) {
			return reconcile.Result{}, nil
```

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: My point is that you use the reconcile method just for the sake of creating namespace:
```go
// first reconcile for ns creation
result, err := r.Reconcile(request)
```
but the reconcile method should be the object of verification, not of the test setup - that should be done via adding the namespace into the fake client.

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: thanks for explaination. Let's see how can we fix it in Che refactoring PR

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: sure, I have fixed it for tekton controller tests and I will fix it for che controller tests in next PR

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Ah, sure, it's the default `openshift-operators` namespace for the AllNamespaces operators. I forgot - thanks for clarifying

File: pkg/controller/tektoninstallation/tektoninstallation_controller_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/tektoninstallation/tektoninstallation_controller_test.go]: I don't understand your answer

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: ok, so test just for the sake of having code-coverage higher... That's what I was mising

File: pkg/controller/tektoninstallation/tektoninstallation_controller_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller_test.go]: We have two tests one is testing controller's reconcile with it's subtests and another is testing non-exported function with different conditions covered in subtests.

I don't understand what's wrong with TestTektonInstallationController

[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/tektoninstallation/tektoninstallation_controller_test.go]: I was referring to the whole file not only single tests, but the file has been either updated or I reviewed outdated version - it looks better now.
But still I don't understand why having `"should reconcile with tekton installation"` and `"should update status when failed to create tekton subscription"` as subtests when they don't share any common setup - they can be as standalone tests - it's easier to read and review.

File: pkg/controller/tektoninstallation/tektoninstallation_controller_test.go [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+
+		t.Run("should create tekton subscription and requeue", func(t *testing.T) {
+			// when
+			_, err := r.Reconcile(request)
+
+			// then
+			require.NoError(t, err)
+
+			AssertThatTektonInstallation(t, tektonInstallation.Namespace, tektonInstallation.Name, cl).
+				HasConditions(tekton.SubscriptionCreated(tekton.SubscriptionSuccess))
```
Comment (nurali-techie) [inline pkg/controller/tektoninstallation/tektoninstallation_controller_test.go]: We can add `AssertThatSubscription()` here too ??

**Comment by nurali-techie** [comment]
@dipak-pawar I have re-viewed tekton part of changes and provided comments .. I guess you parallel fixed them :-)

**Comment by nurali-techie** [comment]
@dipak-pawar I really like the way you organize overall controller code with status/error handling :-)  :+1:

File: pkg/controller/tektoninstallation/tektoninstallation_controller_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller_test.go]: addressed in  7876e23

File: pkg/controller/tektoninstallation/tektoninstallation_controller_test.go (hunk 18 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller_test.go]: Addressed in  40cbf9c

File: pkg/test/toolchain/toolchain.go [hunk 19]
```diff
@@ -22,3 +24,11 @@ func NewCheInstallation(cheNamespace string) *v1alpha1.CheInstallation {
 func GenerateName(prefix string) string {
 	return fmt.Sprintf("%s-%d", prefix, time.Now().UnixNano())
 }
+
+func NewTektonInstallation() *v1alpha1.TektonInstallation {
+	return &v1alpha1.TektonInstallation{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: TektonInstallation,
```
Comment (nurali-techie) [inline pkg/test/toolchain/toolchain.go]: Not sure but is it fine/mandatory to have same name for `TektonInstallation` ??  At least for test, we can create multiple `TektonInstallation` with different name so that if test running parallel then they are not using each other `TektonInstallation` CR .. right ??

File: test/e2e/toolchain_test.go [hunk 20]
```diff
@@ -36,19 +36,18 @@ func TestToolchain(t *testing.T) {
 	tektonSub := tekton.NewSubscription(tekton.SubscriptionNamespace)
 
 	cheInstallation := NewCheInstallation(cheOperatorNs)
+	tektonInstallation := NewTektonInstallation()
+
```
Comment (nurali-techie) [inline test/e2e/toolchain_test.go]: I would prefer two test method `TestCheInstallation()` and `TestTektonInstallation()` instead of one currently `TestToolchain()` .. so adding more test in future will have some segregation.

File: test/wait/awaitility.go [hunk 21]
```diff
@@ -82,26 +93,74 @@ func UntilHasStatusCondition(conditions ...toolchainv1alpha1.Condition) CheInsta
...
+				a.T.Logf("waiting for availability of TektonInstallation '%s'", name)
+				return false, nil
+			}
+			return false, err
+		}
+		for _, isMatched := range waitCond {
+			if !isMatched(a, ti) {
+				return false, nil
+			}
+		}
```
Comment (nurali-techie) [inline test/wait/awaitility.go]: see if you can switch this way of comparing conditions https://github.com/codeready-toolchain/toolchain-e2e/blob/master/wait/member.go#L63-L66

File: test/e2e/toolchain_test.go [hunk 22]
```diff
@@ -63,13 +62,46 @@ func TestToolchain(t *testing.T) {
...
 		AssertThatSubscription(t, tektonSub.Namespace, tektonSub.Name, f.Client).
 			Exists().
 			HasSpec(tektonSub.Spec)
 	})
 
-	t.Run("should remove operatorgroup and subscription for che with installconfig deletion", func(t *testing.T) {
+	t.Run("should recreate deleted subscription for tekton", func(t *testing.T) {
+		// given
+		tektonSubscription, err := await.GetTektonSubscription()
+		require.NoError(t, err)
```
Comment (nurali-techie) [inline test/e2e/toolchain_test.go]: At first sight, waiting for `GetTektonSubscription()` in e2e test looks different.

I was expecting the steps for this test case would be .. 
create_tekton_installation
wait_for_tekton_installation with ready=true
delete_tekton_subscription
wait_for_tekton_subscription to be re-created

but I see this works as we always have one and only one with same name for tekton_subscription .. but still imagine if test are running parallel and `should recreate deleted subscription for tekton` run way before `should create subscription for tekton with TektonInstallation` then is what is currently will work ??  If yes, I am fine.  If no, please do needful changes.

Comment (dipak-pawar) [reply test/e2e/toolchain_test.go]: @nurali-techie we'll be having only one `TektonInstallation` resource as it's cluster scoped. We don't want user to create multiple instances of `TektonInstallation`, and even anyone create it has to be with same name to trigger reconcile. 

Also I don't see how would you run it parallely using `operator-sdk` and not sure why?

File: pkg/test/toolchain/toolchain.go (hunk 19 above)
Comment (dipak-pawar) [reply pkg/test/toolchain/toolchain.go]: No, we need to keep this name same, to have only one resource across cluster scoped

File: test/e2e/toolchain_test.go (hunk 20 above)
Comment (dipak-pawar) [reply test/e2e/toolchain_test.go]: I believe, we are testing end to end, I don't want to seperate it, as we are testing anything randomly and with this way we'll be more confident about it's behaviour

File: test/wait/awaitility.go (hunk 21 above)
Comment (dipak-pawar) [reply test/wait/awaitility.go]: not sure how would it help?

File: test/e2e/toolchain_test.go (hunk 22 above)
Comment (nurali-techie) [reply test/e2e/toolchain_test.go]: ok, make sense :+1:

File: test/wait/awaitility.go (hunk 21 above)
Comment (nurali-techie) [reply test/wait/awaitility.go]: Just a diff_way and I found it simple so suggested but anyway fine with the way it's done currently :+1:

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+func (r *ReconcileTektonInstallation) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling TektonInstallation")
+
+	// Fetch the TektonInstallation instance
+	tektonInstallation := &toolchainv1alpha1.TektonInstallation{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: toolchain.TektonInstallation}, tektonInstallation); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		} else {
```
Comment (MatousJobanek) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: I know that it's just a detail, but you don't need the `else` part here

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+			// Error reading the object - requeue the request.
+			return reconcile.Result{}, err
+		}
+	}
+
+	if err := r.EnsureTektonSubscription(reqLogger, tektonInstallation); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
```
Comment (MatousJobanek) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: same here - you don't need the if statement - you can directly return: `return reconcile.Result{}, err` without checking if the error is nil or not

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+func (r *ReconcileTektonInstallation) EnsureTektonSubscription(logger logr.Logger, tektonInstallation *v1alpha1.TektonInstallation) error {
+	tektonSubNamespace := tekton.SubscriptionNamespace
+	if err := r.ensureTektonSubscription(logger, tektonInstallation, tektonSubNamespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, tektonInstallation, r.setStatusTektonSubscriptionFailed, err, "failed to create tekton subscription in namespace %s", tektonSubNamespace)
+	}
+	return r.StatusUpdate(logger, tektonInstallation, r.setStatusTektonSubscriptionReady, tekton.SubscriptionSuccess)
```
Comment (MatousJobanek) [inline pkg/controller/tektoninstallation/tektoninstallation_controller.go]: I thouth that we don't set any message when the `Condition.status` is `true` - we don't do it in any our CR and I'l love to keep it consistent

File: pkg/test/toolchain/toolchain.go (hunk 19 above)
Comment (MatousJobanek) [reply pkg/test/toolchain/toolchain.go]: > No, we need to keep this name same, to have only one resource across cluster scoped

But that should be applied for the `Cheinstallation` as well, which is not the case as you generate the name there.

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 24 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: sure, this can be done, but when we are introducing `CheCluster`, we need this if statement, so let's keep as it is

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 25 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: @nurali-techie Can you please address this?

File: pkg/test/toolchain/toolchain.go (hunk 19 above)
Comment (dipak-pawar) [reply pkg/test/toolchain/toolchain.go]: sure, it'll come when we are refactoring cheInstallation controller

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 24 above)
Comment (MatousJobanek) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: >but when we are introducing CheCluster, we need this if statement, so let's keep as it is

I don't understand. Why? This is tekton controller, how does it relate to che?

File: pkg/test/toolchain/toolchain.go (hunk 19 above)
Comment (MatousJobanek) [reply pkg/test/toolchain/toolchain.go]: But you already changed that name and the part of the code, so why doing just half of the changes here...?
Anyway, I will trust you that you will change it in the next PR

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 24 above)
Comment (dipak-pawar) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: ohh sorry my bad, I think we ca address it @nurali-techie

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 25 above)
Comment (nurali-techie) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Done https://github.com/codeready-toolchain/toolchain-operator/pull/11/commits/aab0f909c2fa3e5b24e30b3ec06fb5e7516769f3

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 24 above)
Comment (nurali-techie) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Done https://github.com/codeready-toolchain/toolchain-operator/pull/11/commits/aab0f909c2fa3e5b24e30b3ec06fb5e7516769f3

File: pkg/controller/tektoninstallation/tektoninstallation_controller.go (hunk 23 above)
Comment (nurali-techie) [reply pkg/controller/tektoninstallation/tektoninstallation_controller.go]: Done https://github.com/codeready-toolchain/toolchain-operator/pull/11/commits/aab0f909c2fa3e5b24e30b3ec06fb5e7516769f3
